Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformationManager.java

```diff
@@ -17,79 +20,191 @@
+            if (!(subjectNode instanceof PredicateNode)) {
+                link.setSubject(updatedNodes.get(subjectNode.getConceptId()));
+            }
+
+            Node objectNode = link.getObject();
+            if (!(objectNode instanceof PredicateNode)) {
+                link.setObject(updatedNodes.get(objectNode.getConceptId()));
+            }
+        }
+
+        // return new network with updated nodes and links
+        return new TransformedNetwork(updatedNodes, links);
+    }
+
+    @Override
+    public ITransformedNetwork getSearchTransformedNetwork(String projectId, String conceptId)
+        throws QuadrigaStorageException {
+        // get the transformed network and search for the concept id
+        List<INetwork> networkList = getNetworkList(projectId);
+
+        if (networkList == null) {
+            return null;
+        }
+
+        // get the transformed network of all the networks in a project
+        ITransformedNetwork transformedNetwork = getTransformedNetworkOfProject(projectId);
+
+        // Filter the nodes with the concept id
+        // add all the statement ids to a set
+        Set<String> statementIdSearchSet = new HashSet<String>();
+        List<Node> searchedNodes = new ArrayList<Node>();
+        for (Node node: transformedNetwork.getNodes().values()) {
+            if (conceptId.equals(node.getConceptId())) {
+                statementIdSearchSet.addAll(node.getStatementIds());
+            }
+        }
+
+        // include only those links which have statement ids in the search set
+        List<Link> finalLinks = new ArrayList<Link>();
+        for (Link link: transformedNetwork.getLinks()) {
+            if (statementIdSearchSet.contains(link.getStatementId())) {
+                // statement id match
+                // add to the final link list
+                finalLinks.add(link);
+            }
+        }
+
+        // final nodes
+        Map<String, Node> finalNodes = new HashMap<String, Node>();
+        Set<Node> addedNodes = new HashSet<Node>();
+        Integer index = 0;
+        for (Link link: finalLinks) {
+            Node subjectNode = link.getSubject();
+            Node objectNode = link.getObject();
+            // if node already added then do not add it
+            // if nodes are added twice - it would produce many
+            // nodes and less links => disconnected graph
+            if (!addedNodes.contains(subjectNode)) {
+                finalNodes.put((++index).toString(), subjectNode);
+                addedNodes.add(subjectNode);
+            }
+
+            if (!addedNodes.contains(objectNode)) {
+                finalNodes.put((++index).toString(), objectNode);
+                addedNodes.add(objectNode);
+            }
+        }
+
+        // finally add the searched concept nodes if they are not present
+        for (Node node: searchedNodes) {
+            if (!searchedNodes.contains(node)) {
+                finalNodes.put((++index).toString(), node);
```

Comment:
If I do not use another unique id then I have to search the entire map for the key and then add the node with key in the final map. In fact we actually do not use this key anywhere in the program. When you look into `D3Creator.java` class (which would take the nodes and links and create a json out of it) the first statement is 
`List<Node> nodeList = new ArrayList<Node>(nodes.values());`
And the entire code would just use the same `nodeList`

And couple of things regarding `D3Creator.java`
1. It does not use a JSON creator so the json message it constructs would be invalid when the node attributes contain a double quote string.
2. You can see it uses `nodeList.indexOf(link.getObject())` code in couple of places which would search for the entire array list each and every time which is a O(n*n) solution

A better approach would be take the links and iterator over the links. For each link add the subject and object to a list node and we already know the indexes of these two nodes. The time complexity would be O(n).

Please let me know if you think refactoring is necessary. I can work on it.